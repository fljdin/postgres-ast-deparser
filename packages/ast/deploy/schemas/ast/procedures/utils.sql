-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
