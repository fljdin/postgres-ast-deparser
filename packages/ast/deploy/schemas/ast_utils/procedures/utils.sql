-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
