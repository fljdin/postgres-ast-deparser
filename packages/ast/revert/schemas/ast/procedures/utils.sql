-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
