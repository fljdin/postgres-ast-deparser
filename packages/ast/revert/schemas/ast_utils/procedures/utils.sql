-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
