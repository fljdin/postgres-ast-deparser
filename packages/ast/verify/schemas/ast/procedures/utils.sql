-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
