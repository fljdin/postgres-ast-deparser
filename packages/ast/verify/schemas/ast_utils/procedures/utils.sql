-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
