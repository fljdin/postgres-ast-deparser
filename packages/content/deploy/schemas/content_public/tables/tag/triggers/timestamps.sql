-- Deploy schemas/content_public/tables/tag/triggers/timestamps to pg

-- requires: schemas/content_public/schema
-- requires: schemas/content_public/tables/tag/table

BEGIN;

ALTER TABLE content_public.tag ADD COLUMN created_at TIMESTAMPTZ;
ALTER TABLE content_public.tag ALTER COLUMN created_at SET DEFAULT NOW();

ALTER TABLE content_public.tag ADD COLUMN updated_at TIMESTAMPTZ;
ALTER TABLE content_public.tag ALTER COLUMN updated_at SET DEFAULT NOW();

CREATE TRIGGER update_content_public_tags_modtime
BEFORE UPDATE OR INSERT ON content_public.tag
FOR EACH ROW
EXECUTE PROCEDURE tg_update_timestamps();

COMMIT;
