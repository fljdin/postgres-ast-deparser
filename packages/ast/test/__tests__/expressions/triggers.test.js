import { getConnections } from '../../utils';
import { triggers } from './__fixtures__/triggers';

let db, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  await db.begin();
  await db.savepoint();
});
afterAll(async () => {
  try {
    //try catch here allows us to see the sql parsing issues!
    await db.rollback();
    await db.commit();
    await teardown();
  } catch (e) {}
});

it('create trig stmt', async () => {
  const json = triggers[0];
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( '${JSON.stringify(json)}'::jsonb );
  `);
  expect(result).toMatchSnapshot();
});

xit('create_trigger_stmt', async () => {
  const [result] = await db.any(`
select ast.create_trig_stmt(
    v_trigname := 'trigger',
    v_relation := ast.range_var( 
      v_schemaname := 'schema-name',
      v_relname := 'mytable',
      v_inh := true,
      v_relpersistence := 'p'
    ),
    v_funcname := to_jsonb(ARRAY[ ast.string('tg-schema'),ast.string('tgname') ]),
    v_row := true,
    v_timing := 2,
    v_events := 16,
    v_whenClause := ast.a_expr( v_kind := 'AEXPR_DISTINCT', 
        v_lexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('old'),ast.string('field-b') ])
        ),
        v_name := to_jsonb(ARRAY[ast.string('=')]),
        v_rexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('new'),ast.string('field-b') ])
        ) 
    )
);
  `);
  expect(result).toMatchSnapshot();
});

it('create_trigger_stmt deparse', async () => {
  const [result] = await db.any(`
select deparser.deparse( 
  ast.create_trig_stmt(
    v_trigname := 'trigger',
    v_relation := ast.range_var( v_schemaname := 'schema-name', v_relname := 'mytable', v_inh := true, v_relpersistence := 'p'),
    v_funcname := to_jsonb(ARRAY[ ast.string('tg-schema'),ast.string('tgname') ]),
    v_row := true,
    v_timing := 2,
    v_events := 16,
    v_whenClause := ast.a_expr( v_kind := 'AEXPR_DISTINCT', 
        v_lexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('old'),ast.string('field-b') ])
        ),
        v_name := to_jsonb(ARRAY[ast.string('=')]),
        v_rexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('new'),ast.string('field-b') ])
        ) 
    )
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('create_trigger_with_fields', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_trigger_distinct_fields(
    v_trigger_name := 'my-trigger',
    v_schema_name := 'my-schema',
    v_table_name := 'my-table',
    v_trigger_fn_schema := 'my-tg-fn-schema',
    v_trigger_fn_name := 'my-tg-fn',  
    v_fields := ARRAY['name', 'description']::text[],
    v_timing := 2,
    v_events := 4 | 16)
  )`);
  expect(result).toMatchSnapshot();
});

it('create_trigger_with_fields w params', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_trigger_distinct_fields(
    v_trigger_name := 'my-trigger',
    v_schema_name := 'my-schema',
    v_table_name := 'my-table',
    v_trigger_fn_schema := 'my-tg-fn-schema',
    v_trigger_fn_name := 'my-tg-fn',  
    v_fields := ARRAY['name', 'description']::text[],
    v_params := ARRAY['name', 'description']::text[],
    v_timing := 2,
    v_events := 4 | 16)
  )`);
  expect(result).toMatchSnapshot();
});

it('create_trigger_with_fields and names wo quotes', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_trigger_distinct_fields(
    v_trigger_name := 'mytrigger',
    v_schema_name := 'myschema',
    v_table_name := 'mytable',
    v_trigger_fn_schema := 'mytgfnschema',
    v_trigger_fn_name := 'mytgfn',  
    v_fields := ARRAY['name', 'description']::text[],
    v_timing := 2,
    v_events := 4 | 16)
  )`);
  expect(result).toMatchSnapshot();
});

it('ast_helpers.create_trigger_distinct_fields', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_trigger_distinct_fields(
    v_trigger_name := 'mytrigger',
    v_schema_name := 'myschema',
    v_table_name := 'mytable',
    v_trigger_fn_schema := 'mytgfnschema',
    v_trigger_fn_name := 'mytgfn',  
    v_fields := ARRAY['name', 'description', 'hi']::text[],
    v_timing := 2,
    v_events := 4 | 16)
  )`);
  expect(result).toMatchSnapshot();
});

it('ast_helpers.create_trigger_distinct_fields', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_trigger_distinct_fields(
    v_trigger_name := 'mytrigger',
    v_schema_name := 'myschema',
    v_table_name := 'mytable',
    v_trigger_fn_schema := 'mytgfnschema',
    v_trigger_fn_name := 'mytgfn',  
    v_timing := 2,
    v_events := 4 | 16)
  )`);
  expect(result).toMatchSnapshot();
});

it('drop_trigger', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( 
  ast_helpers.drop_trigger(
    v_trigger_name := 'mytrigger',
    v_schema_name := 'myschema',
    v_table_name := 'mytable'
    ))`);
  expect(result).toMatchSnapshot();
});

it('create_function trigger', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_function(
    v_schema_name := 'v_schema_name'::text,
    v_function_name := 'v_function_name'::text,
    v_type := 'TRIGGER'::text,
    v_parameters := NULL,
    v_body := 'RETURN NEW;'::text,
    v_volatility := 'volatile'::text,
    v_language := 'plpgsql'::text,
    v_security := 0::int
  )
  )`);
  expect(result).toMatchSnapshot();
});

it('create_function trigger', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse( 
  ast_helpers.create_function(
    v_schema_name := 'v_schema_name'::text,
    v_function_name := 'v_function_name'::text,
    v_type := 'TRIGGER'::text,
    v_parameters := NULL,
    
    v_body := deparser.deparse(
        ast.raw_stmt(
          v_stmt := ast_helpers.eq(
              v_lexpr := ast.string('NEW.field'),
              v_rexpr := ast_helpers.tsvector_index($1::jsonb)
          ),
          v_stmt_len := 1
    )) || E'\nRETURN NEW;',
    v_volatility := 'volatile'::text,
    v_language := 'plpgsql'::text,
    v_security := 0::int
  )
  )`,
    [
      JSON.stringify([
        {
          lang: 'pg_catalog.simple',
          field: 'NEW.name',
          weight: 'A'
        },
        {
          lang: 'pg_catalog.english',
          field: 'NEW.name',
          weight: 'B'
        },
        {
          lang: 'pg_catalog.english',
          field: 'NEW.tags',
          weight: 'C',
          type: 'citext',
          array: true
        },
        {
          lang: 'pg_catalog.english',
          field: 'NEW.sub_head',
          weight: 'B'
        }
      ])
    ]
  );
  expect(result).toMatchSnapshot();
});

it('create_trigger_stmt deparse', async () => {
  const [result] = await db.any(`
select deparser.deparse( 
  ast.create_trig_stmt(
    v_trigname := 'trigger',
    v_relation := ast.range_var( v_schemaname := 'schema-name', v_relname := 'mytable', v_inh := true, v_relpersistence := 'p'),
    v_funcname := to_jsonb(ARRAY[ ast.string('tg-schema'),ast.string('tgname') ]),
    v_args := to_jsonb(ARRAY[ ast.string('tg-schema'),ast.string('tgname') ]),
    v_row := true,
    v_timing := 2,
    v_events := 16,
    v_whenClause := ast.a_expr( v_kind := 'AEXPR_DISTINCT', 
        v_lexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('old'),ast.string('field-b') ])
        ),
        v_name := to_jsonb(ARRAY[ast.string('=')]),
        v_rexpr := ast.column_ref(
          to_jsonb(ARRAY[ ast.string('new'),ast.string('field-b') ])
        ) 
    )
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('create_trigger_distinct_fields deparse', async () => {
  const [result] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_trigger_distinct_fields(
    v_trigger_name := 'v_trigger_name',
    
    v_schema_name := 'v_schema_name',
    v_table_name := 'v_table_name',

    v_trigger_fn_schema := 'v_trigger_fn_schema',
    v_trigger_fn_name := 'v_trigger_fn_name',

    v_fields := ARRAY['field1', 'field2'],
    v_params := NULL,
    v_timing := 2,
    v_events := 4
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('create_trigger deparse', async () => {
  const [result] = await db.any(`
select deparser.deparse( 
  ast_helpers.create_trigger(
    v_trigger_name := 'v_trigger_name',
    
    v_schema_name := 'v_schema_name',
    v_table_name := 'v_table_name',

    v_trigger_fn_schema := 'v_trigger_fn_schema',
    v_trigger_fn_name := 'v_trigger_fn_name',

    v_whenClause := ast_helpers.neq(
      v_lexpr := ast_helpers.col('new', 'type'),
      v_rexpr := ast.a_const( v_val := ast.integer(0) )
    ),
    v_params := NULL,
    v_timing := 2,
    v_events := 4
  )
);
  `);
  expect(result).toMatchSnapshot();
});
