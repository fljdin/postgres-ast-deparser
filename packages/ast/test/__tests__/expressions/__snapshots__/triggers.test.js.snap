// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create trig stmt 1`] = `
"CREATE TRIGGER trigger_name 
 BEFORE UPDATE ON \\"schema-name\\".table_name 
 FOR EACH ROW 
 WHEN (NEW.field_name IS DISTINCT FROM OLD.field_name) 
 EXECUTE PROCEDURE trigger_function_schema.trigger_function ( )"
`;

exports[`createtrigger with fields 1`] = `
"CREATE TRIGGER \\"my-trigger\\" 
 BEFORE INSERT OR UPDATE ON \\"my-schema\\".\\"my-table\\" 
 FOR EACH ROW 
 WHEN ((OLD.description IS DISTINCT FROM NEW.description OR OLD.name IS DISTINCT FROM NEW.name)) 
 EXECUTE PROCEDURE \\"my-tg-fn-schema\\".\\"my-tg-fn\\" ( )"
`;

exports[`createtrigger with fields and names wo quotes 1`] = `
"CREATE TRIGGER mytrigger 
 BEFORE INSERT OR UPDATE ON myschema.mytable 
 FOR EACH ROW 
 WHEN ((OLD.description IS DISTINCT FROM NEW.description OR OLD.name IS DISTINCT FROM NEW.name)) 
 EXECUTE PROCEDURE mytgfnschema.mytgfn ( )"
`;

exports[`createtriggerstmt 1`] = `
Object {
  "createtriggerstmt": Object {
    "CreateTrigStmt": Object {
      "events": 16,
      "funcname": Array [
        Object {
          "String": Object {
            "str": "tg-schema",
          },
        },
        Object {
          "String": Object {
            "str": "tgname",
          },
        },
      ],
      "relation": Object {
        "RangeVar": Object {
          "inh": true,
          "relname": "mytable",
          "relpersistence": "p",
          "schemaname": "schema-name",
        },
      },
      "row": true,
      "timing": 2,
      "trigname": "trigger",
      "whenClause": Object {
        "A_Expr": Object {
          "kind": 3,
          "lexpr": Object {
            "ColumnRef": Object {
              "fields": Array [
                Object {
                  "String": Object {
                    "str": "old",
                  },
                },
                Object {
                  "String": Object {
                    "str": "field-a",
                  },
                },
              ],
            },
          },
          "name": Array [
            Object {
              "String": Object {
                "str": "=",
              },
            },
          ],
          "rexpr": Object {
            "ColumnRef": Object {
              "fields": Array [
                Object {
                  "String": Object {
                    "str": "new",
                  },
                },
                Object {
                  "String": Object {
                    "str": "field-a",
                  },
                },
              ],
            },
          },
        },
      },
    },
  },
}
`;

exports[`createtriggerstmt deparse 1`] = `
Object {
  "deparse": "CREATE TRIGGER trigger 
 BEFORE UPDATE ON \\"schema-name\\".mytable 
 FOR EACH ROW 
 WHEN (OLD.\\"field-a\\" IS DISTINCT FROM NEW.\\"field-a\\") 
 EXECUTE PROCEDURE \\"tg-schema\\".tgname ( )",
}
`;
